pkg/webhook/admission/pod/storage_initializer_injector.go

```diff
@@ -481,16 +481,127 @@ func (mi *StorageInitializerInjector) InjectStorageInitializer(pod *v1.Pod) erro
+		if c.Name == StorageInitializerContainerName {
+			storageInitializerContainer = &pod.Spec.InitContainers[idx]
+		}
+	}
+
+	// If the storage initializer is not injected, there is no action to do
+	if storageInitializerContainer == nil {
+		return nil
+	}
+
 	// Allow to override the uid for the case where ISTIO CNI with DNS proxy is enabled
 	// See for more: https://istio.io/latest/docs/setup/additional-setup/cni/#compatibility-with-application-init-containers.
 	if value, ok := pod.GetAnnotations()[constants.IstioSidecarUIDAnnotationKey]; ok {
 		if uid, err := strconv.ParseInt(value, 10, 64); err == nil {
-			initContainer.SecurityContext.RunAsUser = ptr.Int64(uid)
+			if storageInitializerContainer.SecurityContext == nil {
+				storageInitializerContainer.SecurityContext = &v1.SecurityContext{}
+			}
+			storageInitializerContainer.SecurityContext.RunAsUser = ptr.Int64(uid)
+		}
+	} else {
+		// When Istio CNI is disabled, the istio-init container would be present.
+		// If it is there, there is no need to touch the security context of the pod.
+		// Reference: https://github.com/istio/istio/blob/d533e52acc54b4721d23b1332aea1f234ecbe3e6/pkg/config/analysis/analyzers/maturity/maturity.go#L134
+		for _, container := range pod.Spec.InitContainers {
+			if container.Name == constants.IstioInitContainerName {
+				return nil
+			}
 		}
-	}
 
-	// Add init container to the spec
-	pod.Spec.InitContainers = append(pod.Spec.InitContainers, *initContainer)
+		// When Istio CNI is enabled, a sidecar.istio.io/interceptionMode annotation is injected to the pod.
+		// There are three interception modes: REDIRECT, TPROXY and NONE.
+		// It only makes sense to adjust the security context of the storage initializer if REDIRECT mode is
+		// observed, because the Istio sidecar would be injected and traffic would be sent to it, but the
+		// sidecar won't be running at PodInitialization phase.
+		// The TPROXY mode can indicate that Istio Ambient is enabled. The Waypoint proxy would already be running and
+		// captured traffic can go through.
+		// The TPROXY mode can also be set by the user. If this is the case, it is not possible to infer the setup.
+		// Lastly, if interception mode is NONE the traffic is not being captured. This is an advanced mode, and
+		// it is not possible to infer the setup.
+		istioInterceptionMode := pod.Annotations[constants.IstioInterceptionModeAnnotation]
+		if istioInterceptionMode != constants.IstioInterceptModeRedirect {
+			return nil
+		}
+
+		// The storage initializer can only run smoothly when running with the same UID as the Istio sidecar.
+		// First, find the name of the Istio sidecar container. This is found in a status annotation injected
+		// by Istio. If there is no Istio sidecar status annotation, assume that the pod does
+		// not have a sidecar and leave untouched the security context.
+		istioStatus, istioStatusOk := pod.Annotations[constants.IstioSidecarStatusAnnotation]
```

Comment:
Is this true for all the supported Istio versions ?

Reply:
Yes, this annotation has been available for quite a while.